subxt/src/rpc/lightclient/background.rs

```diff
@@ -0,0 +1,275 @@
+use futures::stream::StreamExt;
+use futures_util::future::{self, Either};
+use serde::Deserialize;
+use serde_json::value::RawValue;
+use std::{collections::HashMap, str::FromStr};
+use tokio::sync::{mpsc, oneshot};
+
+///! The background task of the light client.
+
+/// The number of notifications that are buffered, before the user
+/// registers to [`LightClientInner::register_subscription`].
+/// Only the first `BUFFER_NUM_NOTIFICATIONS` are buffered, while the
+/// others are ignored.
+///
+/// In practice, the Light Client produces notifications at a lower rate
+/// than the substrate full node.
+const BUFFER_NUM_NOTIFICATIONS: usize = 16;
+
+const LOG_TARGET: &str = "light-client-background";
+
+/// Message protocol between the front-end client that submits the RPC requests
+/// and the backend handler that produces responses from the chain.
+///
+/// The light client uses a single object [`smoldot_light::JsonRpcResponses`] to
+/// handle all requests and subscriptions from a chain. A background task is spawned
+/// to multiplex the rpc responses and to provide them back to their rightful submitters.
+///
+/// This presumes that the request ID for both method calls and subscriptions is unique.
+#[derive(Debug)]
+pub enum BackendMessage {
+    /// The RPC method request.
+    Request {
+        /// ID of the request, needed to match the result.
+        id: String,
+        /// Channel used to send back the result.
+        sender: oneshot::Sender<Box<RawValue>>,
+    },
+    /// The RPC subscription (pub/sub) request.
+    Subscription {
+        /// ID of the subscription, needed to match the result.
+        id: String,
+        /// Channel used to send back the notifcations.
+        sender: mpsc::Sender<Box<RawValue>>,
+    },
+}
+
+/// Background task data.
+#[derive(Default)]
+pub struct BackgroundTask {
+    /// Map the request ID of a RPC method to the frontend `Sender`.
+    requests: HashMap<String, oneshot::Sender<Box<RawValue>>>,
+    /// Map the subscription ID to the frontend `Sender`.
+    subscriptions: HashMap<String, mpsc::Sender<Box<RawValue>>>,
+    /// Notifications are cached for users that did not subscribe yet.
+    subscriptions_cache: HashMap<String, Vec<Box<RawValue>>>,
+}
+
+impl BackgroundTask {
+    /// Constructs a new [`BackgroundTask`].
+    pub fn new() -> BackgroundTask {
+        BackgroundTask::default()
+    }
+
+    /// Handle the registration messages received from the user.
+    async fn handle_register(&mut self, message: BackendMessage) {
+        match message {
+            BackendMessage::Request { id, sender } => {
+                self.requests.insert(id, sender);
+            }
+            BackendMessage::Subscription { id, sender } => {
+                // Drain the subscription cache, that holds messages that
```

Comment:
I wonder if this is needed anymore, IIRC this was a bug in the old jsonrpc server

fine let's keep this for now